File: src/algorithms/linked-lists/traversal/find-node-by-value/find-node-by-value.test.ts
import { describe, it, expect } from "vitest";
import { findNodeByValue } from "./sources/find-node-by-value.ts?fn";

interface ListNode {
  value: number;
  next: ListNode | null;
}

function buildList(values: number[]): ListNode | null {
  let head: ListNode | null = null;
  for (let idx = values.length - 1; idx >= 0; idx--) {
    head = { value: values[idx]!, next: head };
  }
  return head;
}

describe("findNodeByValue", () => {
  it("returns the node when target is found at head", () => {
    const result = findNodeByValue(buildList([5, 2, 3, 4]), 5);
    expect(result).not.toBeNull();
    expect((result as ListNode).value).toBe(5);
  });

  it("returns the node when target is found in the middle", () => {
    const result = findNodeByValue(buildList([1, 2, 7, 4, 5]), 7);
    expect(result).not.toBeNull();
    expect((result as ListNode).value).toBe(7);
  });

  it("returns the node when target is found at the end", () => {
    const result = findNodeByValue(buildList([1, 2, 3, 9]), 9);
    expect(result).not.toBeNull();
    expect((result as ListNode).value).toBe(9);
  });

  it("returns null when target is not found", () => {
    const result = findNodeByValue(buildList([1, 2, 3, 4]), 42);
    expect(result).toBeNull();
  });

  it("returns null for an empty list", () => {
    const result = findNodeByValue(null, 5);
    expect(result).toBeNull();
  });

  it("returns the node for single-node list when target matches", () => {
    const result = findNodeByValue(buildList([42]), 42);
    expect(result).not.toBeNull();
    expect((result as ListNode).value).toBe(42);
  });

  it("returns null for single-node list when target does not match", () => {
    const result = findNodeByValue(buildList([42]), 7);
    expect(result).toBeNull();
  });
});

File: src/algorithms/linked-lists/traversal/find-node-by-value/sources/FindNodeByValue_test.cpp
#include <cassert>
#include <vector>
#include "FindNodeByValue.cpp"

ListNode* buildList(const std::vector<int>& values) {
    ListNode* head = nullptr;
    for (int idx = static_cast<int>(values.size()) - 1; idx >= 0; idx--) {
        ListNode* node = new ListNode(values[idx]);
        node->next = head;
        head = node;
    }
    return head;
}

int main() {
    // returns the node when target is found at head
    ListNode* resultAtHead = findNodeByValue(buildList({5, 2, 3, 4}), 5);
    assert(resultAtHead != nullptr && resultAtHead->value == 5);

    // returns the node when target is found in the middle
    ListNode* resultInMiddle = findNodeByValue(buildList({1, 2, 7, 4, 5}), 7);
    assert(resultInMiddle != nullptr && resultInMiddle->value == 7);

    // returns the node when target is found at the end
    ListNode* resultAtEnd = findNodeByValue(buildList({1, 2, 3, 9}), 9);
    assert(resultAtEnd != nullptr && resultAtEnd->value == 9);

    // returns nullptr when target is not found
    assert(findNodeByValue(buildList({1, 2, 3, 4}), 42) == nullptr);

    // returns nullptr for an empty list
    assert(findNodeByValue(nullptr, 5) == nullptr);

    // returns the node for single-node list when target matches
    ListNode* resultSingleMatch = findNodeByValue(buildList({42}), 42);
    assert(resultSingleMatch != nullptr && resultSingleMatch->value == 42);

    // returns nullptr for single-node list when target does not match
    assert(findNodeByValue(buildList({42}), 7) == nullptr);

    return 0;
}
